### PR TITLE
expression: ConstItem => ConstLevel to provide more clear semantics

### LINKED ISSUE #49806: ConstItem => ConstLevel to provide more clear semantics

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #49806

### What changed and how does it work?

Provide `ConstLevel` for `Expression` for better semantics

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/expression/expression.go [hunk 1]
```diff
@@ -84,6 +84,20 @@ type TraverseAction interface {
...
+
+const (
+	// ConstNone indicates the expression is not a constant expression.
+	// The evaluation result may be different for different input rows.
+	ConstNone ConstLevel = iota
+	// ConstOnlyInContext indicates the expression is only a constant for a same context.
+	ConstOnlyInContext
+	// ConstStrict indicates the expression is a constant expression.
+	// The evaluation result is always the same no matter the input context or rows.
+	ConstStrict
```
Comment (qw4990) [inline pkg/expression/expression.go:98]:
```suggestion
	// ConstNone indicates the expression is not a constant expression.
	// The evaluation result may be different for different input rows.
	// e.g. `col_a * 2`, `substring(col_b, 5, 3)`.
	ConstNone ConstLevel = iota
	// ConstOnlyInContext indicates the expression is only a constant for a same context.
	// This is mainly for Plan Cache, e.g. `prepare st from 'select * from t where a<1+?'`, where 
	// the value of `?` may change between different Context.
	ConstOnlyInContext
	// ConstStrict indicates the expression is a constant expression.
	// The evaluation result is always the same no matter the input context or rows.
	// e.g. `1 + 2`, `substring("TiDB SQL Tutorial", 5, 3) + 'abcde'`
	ConstStrict
```

File: pkg/planner/core/rule_partition_processor.go [hunk 2]
```diff
@@ -1500,7 +1500,13 @@ func (p *rangePruner) extractDataForPrune(sctx sessionctx.Context, expr expressi
...
 	// Suppose the partition expression is 'a + b' and we have a condition 'a = 2',
 	// the constExpr is '2 + b' after the replacement which we can't evaluate.
-	if !constExpr.ConstItem(sctx.GetSessionVars().StmtCtx.UseCache) {
+	constLevel := constExpr.ConstLevel()
+	if constLevel == expression.ConstNone {
+		return ret, false
+	}
+	if sctx.GetSessionVars().StmtCtx.UseCache && constLevel != expression.ConstStrict {
+		// When use cache, expression will be shared across contexts.
+		// We should make sure it is always an constant
```
Comment (qw4990) [inline pkg/planner/core/rule_partition_processor.go:1509]: Can we wrap this as a separate function, like `ConstForPlanCache(sctx, expr)`?

**Review by qw4990 — APPROVED** [review]

**Review by YangKeao — APPROVED** [review]
LGTM

File: pkg/planner/core/scalar_subq_expression.go [hunk 3]
```diff
@@ -178,9 +178,9 @@ func (*ScalarSubQueryExpr) IsCorrelated() bool {
 	return false
 }
 
-// ConstItem implements the Expression interface.
-func (*ScalarSubQueryExpr) ConstItem(_ bool) bool {
-	return true
+// ConstLevel returns the const level for the expression
+func (*ScalarSubQueryExpr) ConstLevel() expression.ConstLevel {
+	return expression.ConstNone
```
Comment (windtalker) [inline pkg/planner/core/scalar_subq_expression.go:183]: Looks like the original logic is return ConstStrict here?

Comment (lcwangchao) [reply pkg/planner/core/scalar_subq_expression.go:183]: Yes， confirmed with @qw4990 , returning `ConstNone` instead `ConstStrict` should be more safe

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**